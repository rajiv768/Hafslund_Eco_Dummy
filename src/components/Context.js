import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';

import loadLayers from '../utils/loadLayers';

const defaultContext = {
  onOpen() {},
  onClose() {},
};

export const WidgetContext = React.createContext(defaultContext);

export const WidgetProvider = ({ wab, esriJS, onOpen, onClose, children }) => {
  // set state
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);
  const [layers, setLayers] = useState(null);

  // load layers
  const loadMapLyrs = async () => {
    try {
      const lyrs = await loadLayers(
        esriJS,
        wab.map,
        wab.config.layerCollection,
      );
      setLayers(lyrs);
      setLoading(false);
    } catch (err) {
      setLoading(false);
      setError(true);
    }
  };

  useEffect(() => {
    loadMapLyrs();
  }, []);

  return (
    <WidgetContext.Provider
      value={{
        wab,
        esriJS,
        onOpen,
        onClose,
        loading,
        error,
        layers,
        version: process.env.VERSION || null,
      }}
    >
      {children}
    </WidgetContext.Provider>
  );
};

WidgetProvider.propTypes = {
  wab: PropTypes.object,
  esriJS: PropTypes.object,
  onOpen: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
  children: PropTypes.node,
};

WidgetProvider.defaultProps = {
  wab: {},
  esriJS: {},
  children: null,
};
