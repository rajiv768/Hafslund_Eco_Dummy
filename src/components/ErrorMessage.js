import React from 'react';

const ErrorMessage = () => {
  return (
    <div>
      <p>Sorry â€” something has gone wrong.</p>
    </div>
  );
};

export default ErrorMessage;
