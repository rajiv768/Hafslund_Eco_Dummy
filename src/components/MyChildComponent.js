import React from 'react';
import PropTypes from 'prop-types';
import { Segment } from 'semantic-ui-react';

const MyChildComponent = ({ msg, open }) => {
  return (
    <Segment>
      <h3>{msg}</h3>
      {open && <h5>Widget Open</h5>}
    </Segment>
  );
};

MyChildComponent.propTypes = {
  msg: PropTypes.string,
  open: PropTypes.bool,
};

MyChildComponent.defaultProps = {
  msg: 'Hello World',
  open: true,
};

export default MyChildComponent;
