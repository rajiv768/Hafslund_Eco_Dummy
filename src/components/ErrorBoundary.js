import React, { Component } from 'react';
import PropTypes from 'prop-types';

import ErrorMessage from './ErrorMessage';

class ErrorBoundary extends Component {
  state = {
    error: null,
    errorInfo: null,
  };

  componentDidCatch(error, errorInfo) {
    console.error(error);
    console.error(errorInfo);
    this.setState({ error, errorInfo });
  }

  render() {
    if (this.state.error) {
      return <ErrorMessage />;
    }
    return this.props.children;
  }
}

ErrorBoundary.propTypes = {
  children: PropTypes.element.isRequired,
};

export default ErrorBoundary;
