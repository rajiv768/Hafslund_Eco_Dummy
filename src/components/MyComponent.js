/* eslint-disable no-console */
import React, { useEffect, useState, useContext } from 'react';
import PropTypes from 'prop-types';
import { Header } from 'semantic-ui-react';

import { WidgetContext } from './Context';
import MyChildComponent from './MyChildComponent';

const MyComponent = ({ title }) => {
  const context = useContext(WidgetContext);
  // context object will contain:
  /*
    {
      wab,
      esriJS,
      onOpen,
      onClose,
      loading,
      error,
      layers,
      version
    }
  */
  // console.log('Widget Context:', context);

  // set state
  const [msg, setMsg] = useState('WAB');
  const [open, setOpen] = useState(true);

  // set message
  const updateMsg = newMsg => {
    setTimeout(() => {
      setMsg(newMsg);
    }, 1000);
  };

  // handle widget open and close
  const handleOpen = () => {
    console.log('Open widget');
    setOpen(true);
  };

  const handleClose = () => {
    console.log('Close widget');
    setOpen(false);
  };

  // lifecycle
  useEffect(() => {
    // init open/close handlers
    context.onOpen(handleOpen);
    context.onClose(handleClose);
    // set msg
    updateMsg('WAB React Widget Template');
  }, []);

  if (context.loading) {
    return <div>Loading...</div>;
  }

  if (context.error) {
    return <div>Error...</div>;
  }

  return (
    <div css="padding: 0.25rem;">
      <Header>{title}</Header>
      <MyChildComponent msg={msg} open={open} />
      <div css="font-size: 0.7rem; color: #999999ab;">
        Version: {context.version}
      </div>
    </div>
  );
};

MyComponent.propTypes = {
  title: PropTypes.string,
};

MyComponent.defaultProps = {
  title: 'Hello',
};

export default MyComponent;
