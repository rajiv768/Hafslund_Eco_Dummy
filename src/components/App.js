import 'core-js/stable';
import 'regenerator-runtime/runtime';

import React from 'react';
import PropTypes from 'prop-types';

import ErrorBoundary from './ErrorBoundary';
import { WidgetProvider } from './Context';
import MyComponent from './MyComponent';

const App = props => {
  return (
    <ErrorBoundary>
      <WidgetProvider
        wab={props.wab}
        esriJS={props.esriJS}
        onOpen={props.onOpen}
        onClose={props.onClose}
      >
        <MyComponent title="WAB Widget" />
      </WidgetProvider>
    </ErrorBoundary>
  );
};

App.propTypes = {
  wab: PropTypes.object,
  esriJS: PropTypes.object,
  onOpen: PropTypes.func.isRequired,
  onClose: PropTypes.func.isRequired,
};

App.defaultProps = {
  wab: {},
  esriJS: {},
};

export default App;
