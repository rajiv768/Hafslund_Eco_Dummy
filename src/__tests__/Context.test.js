import React from 'react';
import { render, wait, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import { WidgetProvider } from '../components/Context';

import mock from './__mock__/propsMock';

afterEach(cleanup);

describe('<WidgetProvider />', () => {
  it('renders the widget', async () => {
    // render component
    const { container } = render(
      <WidgetProvider
        wab={mock.wab}
        esriJS={mock.esriJS}
        onOpen={jest.fn}
        onClose={jest.fn}
      >
        <div>children</div>
      </WidgetProvider>,
    );

    // wait for useEffect
    await wait(() => expect(container.firstChild).toMatchSnapshot());
  });
});
