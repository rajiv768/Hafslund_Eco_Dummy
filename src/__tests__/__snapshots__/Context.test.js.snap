// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<WidgetProvider /> renders the widget 1`] = `
<div>
  children
</div>
`;
